torch/nn/attention/_flex_attention.py

```diff
@@ -346,20 +433,74 @@ def _create_mask(
     return mask
 
 
+def _create_mask_from_mask_fn(
+    mask_fn: _mask_fn_signature,
+    B: int,
+    H: int,
+    M: int,
+    N: int,
+    device: str = "cuda",
+    _compiled: bool = False,
+) -> torch.Tensor:
+    r"""This function creates a mask tensor from a score_mod function.
+    Args:
+        mask_fn (Callable): Mask function.
+        B (int): Batch size.
+        H (int): Number of heads.
+        M (int): Sequence length of query.
+        N (int): Sequence length of key/value.
+        device (str): Device to run the mask creation on.
+    Returns:
+        mask (Tensor): A mask tensor with shape (B, H, M, N).
+    """
+    b = torch.arange(0, B, device=device)
+    h = torch.arange(0, H, device=device)
+    m = torch.arange(0, M, device=device)
+    n = torch.arange(0, N, device=device)
+    # TODO: fix this
+    # A hack required because of lack of torchfunctionmode support
+    # Working around some bugs with compiling vmap
+    if _compiled:
+        ctx = nullcontext()
+    else:
+        ctx = TransformGetItemToIndex()  # type: ignore[assignment]
+    mask_fn = torch.vmap(mask_fn, in_dims=(None, None, None, 0))
+    mask_fn = torch.vmap(mask_fn, in_dims=(None, None, 0, None))
+    mask_fn = torch.vmap(mask_fn, in_dims=(None, 0, None, None))
+    mask_fn = torch.vmap(mask_fn, in_dims=(0, None, None, None))
+    with ctx:
+        mask = mask_fn(b, h, m, n)
+    return mask
+
+
 # Done as a workaround around torch.compile not compiling what we want in the
 # presence of the torchfunctionmdoe
 def _create_block_mask_inner(
-    score_mod, B, H, M, N, device, KV_BLOCK_SIZE, Q_BLOCK_SIZE
+    fn, B, H, M, N, device, KV_BLOCK_SIZE, Q_BLOCK_SIZE, is_score_mod
 ):
-    mask = _create_mask(score_mod, B, H, M, N, device, _compiled=True)
-    block_mask = _convert_mask_to_block_mask(
-        mask, KV_BLOCK_SIZE=KV_BLOCK_SIZE, Q_BLOCK_SIZE=Q_BLOCK_SIZE
+    if is_score_mod:
+        # fn is a score_mod function
+        mask_fn = None
+        mask_tensor = _create_mask_from_score_mod(
+            fn, B, H, M, N, device, _compiled=True
```

Comment:
I'm not sure this `_compiled` propagation logic is right... but I'm not sure it was right before either.